Re-use RenderStyle between MediaQuery evaluations.

mstyura opened the pull request:
#### 9213cf0bfaf50b9e6a53de8948d669fe1f770fe3
<pre>
Re-use RenderStyle between MediaQuery evaluations.

<a href="https://bugs.webkit.org/show_bug.cgi?id=293538">https://bugs.webkit.org/show_bug.cgi?id=293538</a>

Reviewed by NOBODY (OOPS!).

When `Page::updateRendering()` is called repeatedly (e.g., 60 times
per second), and the `Document` has many associated
`window.matchMedia(...)` instances, a significant amount of CPU time is
spent constructing and destructing `RenderStyle` instances.

Since the `RenderStyle` used for evaluating media queries only depends
on the `Document` settings and quirks mode, it can be cached and reused
between evaluations, avoiding unnecessary object creation and
improving performance.

* Source/WebCore/css/query/MediaQueryEvaluator.cpp:
(WebCore::MQ::MediaQueryEvaluator::evaluate const):
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::invalidateMediaQueryDefaultStyle):
(WebCore::Document::mediaQueryDefaultStyle const):
(WebCore::Document::setCompatibilityMode):
* Source/WebCore/dom/Document.h:
* Source/WebCore/dom/Page.cpp:
(WebCore::Page::setNeedsRecalcStyleInAllFrames):
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9213cf0bfaf50b9e6a53de8948d669fe1f770fe3

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/105599 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/25348 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/15740 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/110796 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/56210 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/107640 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/25799 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/33849 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/80199 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/108605 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/20146 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/95297 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/60508 "Passed tests") | 
| | [💥 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/19893 "An unexpected error occured. Recent messages:Printed configuration") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/13389 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/55634 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/89595 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/13428 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/113628 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/32737 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/24135 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/89276 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/33099 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/91526 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/88940 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/33808 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/11610 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/28179 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/32663 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/38047 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/32409 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/35757 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/34006 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mdubet commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -1327,6 +1329,22 @@ MediaQueryMatcher& Document::mediaQueryMatcher()
     return *m_mediaQueryMatcher;
 }
 
+const RenderStyle* Document::mediaQueryDefaultStyle() const
```
This should always return a valid RenderStyle so it should return a `const RenderStyle&` instead

darinadler replied:
Agreed:
```suggestion
const RenderStyle& Document::mediaQueryDefaultStyle() const
```

mdubet reviewed (approved):
Great patch ! We could convert the example to a PerformanceTest.

It seems that the RenderStyle for MQ evaluation only depends on `settings.defaultFixedFontSize` and `settings.defaultFontSize` indeed.

darinadler commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -1327,6 +1329,22 @@ MediaQueryMatcher& Document::mediaQueryMatcher()
 
+const RenderStyle* Document::mediaQueryDefaultStyle() const
+{
+    if (m_cachedMediaQueryDefaultStyle == nullptr) {
```
WebKit coding style:
```suggestion
    if (!m_cachedMediaQueryDefaultStyle)
```

darinadler commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -1327,6 +1329,22 @@ MediaQueryMatcher& Document::mediaQueryMatcher()
+const RenderStyle* Document::mediaQueryDefaultStyle() const
+{
+    if (m_cachedMediaQueryDefaultStyle == nullptr) {
+        auto defaultStyle = RenderStyle::createPtr();
```
Another WebKit coding style issue. This isn’t my personal preference, but it’s what we’re doing as a project. Recently, to make lifetime clearer, we have been writing out smart pointer types rather than using `auto`. No need to be explicit about the type of the pointed-to object, but we are explicit about the type of pointer. So we write `Ref` or `RefPtr` or whatever.
```suggestion
        std::unique_ptr defaultStyle = RenderStyle::createPtr();
```

darinadler commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -1327,6 +1329,22 @@ MediaQueryMatcher& Document::mediaQueryMatcher()
+        }
+        m_cachedMediaQueryDefaultStyle = WTFMove(defaultStyle);
+    }
+    return m_cachedMediaQueryDefaultStyle.get();
```
```suggestion
    return *m_cachedMediaQueryDefaultStyle;
```

darinadler commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -1335,6 +1353,7 @@ void Document::setCompatibilityMode(DocumentCompatibilityMode mode)
     m_compatibilityMode = mode;
 
     if (inQuirksMode() != wasInQuirksMode) {
+        m_cachedMediaQueryDefaultStyle.reset();
```
This doesn’t seem like sufficient invalidation. The value returned from `Style::fontSizeForKeyword` when passed `false` depends on the document’s setting’s `defaultFontSize`, so we also need to invalidate when the settings on the document are changed. While `Document::m_settings` can’t be reassigned, the values in the settings can be changed. To handle that properly, I think this needs to be invalidated, probably by `Style::Scope::didChangeStyleSheetEnvironment`. We should also consider how to test that it works properly. I suspect we can make a test that would use an internals mechanism and would demonstrate the cache doing the wrong thing.

darinadler replied:
Please let me know if I’m mistaken about this; I’d be happy to be proven wrong

darinadler replied:
We could also consider naming this in a way that makes it clear that it’s a style that’s dependent only on quirks mode and font size, not more broadly on other aspects of the state of the document. I don’t immediately have a name in mind, though!

mstyura replied:
> While Document::m_settings can’t be reassigned, the values in the settings can be changed.
...
Please let me know if I’m mistaken about this; I’d be happy to be proven wrong

~~Here is the simplified way I've initially understood relationship between `Document` and it's `Settings`:~~
<details>
<summary>snippet</summary>

``` cpp
// auto-generated from https://github.com/WebKit/WebKit/blob/main/Source/WebCore/Scripts/SettingsTemplates/Settings.h.erb
// no members annotated has mutable annotations, so as far as I understsand it means
// const methods guaranteed not to change any object's fields;
struct SettingValues {
  int32_t value;
};

class Settings {
public:
  Settings(): m_setting({.value = 0}) {
  }
  
  void update_value(int32_t value) {
    m_setting.value = value;
  }

private:
  SettingValues m_setting;
};

class Document {

public:
  Document(const Settings& settings): m_settings(settings) {
    
  }

  void do_action() {
    // Document itself can't anyhow change it's own settings via const refenrece it has and taking into account
    // nor SettingValues nor Settings has no mutable annotation on any of fields.

    // 'this' argument to member function 'update_value' has type 'const Settings', but function is not marked const
    // m_settings.update_value(5);
  }
  
private:
  const Settings m_settings;
};
```
</details>

~~In that understanding there is no way to document to mutate it's settings. The only potential problem is that actually `Settings` defined the following way:~~
<details>
<summary>snippet</summary>

```c++
class Settings : public SettingsBase, public RefCounted<Settings> {
    WTF_MAKE_NONCOPYABLE(Settings); WTF_MAKE_FAST_ALLOCATED;
```
</details>

~~and what does happen here `m_settings(settings)` was initially unclear to me, but after debugging and looking into assembly there is just incrementing ref count to reference counted object. So the "owner" of mutable reference to `Settings` instance indeed may unnoticeably modify any of the settings (potentially, I only found the webrtc codecs releated flags changed so far).~~

~~The quick search over document did not reveal any explicit mechanism to handle settings change after document created. How much other computations are obsolete in case of settings change is unclear.~~

~~Currently it seems infeasible (at least for me in reasonable time) to properly implement handling of settings change (it's not clear if any of them changeable at all under normal operation).~~

~~But I see there could be a several compromise solutions here:~~
1. ~~Reset cache just before `m_mediaQueryMatcher->evaluateAll` which could eliminate a lot of redundant computations already, but it'a a bit awkward control flow to read RenderStyle back from document from several calls inside of `evaluateAll`;~~
2. ~~Change signature of `m_mediaQueryMatcher->evaluateAll` to accept `RenderStyle` which will be computed each time before invocation `evaluateAll`, limiting frequency of `RenderStyle` creation to number of `Page::updateRendering` calls, but will not grow linearly to the number of media query lists.~~

~~Do you think those options might make sense to relatively easy get performance improvements due to simplicity of change, but without producing new caching invalidation issue?~~

mstyura replied:
@darinadler Thank you very much for the review! Please disregard my previous comment - it was mostly incorrect and misleading. I was tired and in a hurry yesterday, so I didn't fully understand your feedback.

Now, I believe I've managed to implement what you were suggesting.
The key realization for me was that whenever any value inside `Settings` changes, `Page::setNeedsRelayoutAllFrames` is invoked, and each document is iterated and notified about the change. That seems like the ideal place to invalidate caches that depend on document settings.
Since the code inside `Settings` is auto-generated, it appears to reliably notify about changes.

darinadler reviewed (changes requested):
This seems like a good optimization and I’m really glad we’re pursing it. I don’t _think_ it’s correct yet.

mstyura committed:
Re-use RenderStyle between MediaQuery evaluations.

https://bugs.webkit.org/show_bug.cgi?id=293538

Reviewed by NOBODY (OOPS!).

When `Page::updateRendering()` is called repeatedly (e.g., 60 times
per second), and the `Document` has many associated
`window.matchMedia(...)` instances, a significant amount of CPU time is
spent constructing and destructing `RenderStyle` instances.

Since the `RenderStyle` used for evaluating media queries only depends
on the `Document` settings and quirks mode, it can be cached and reused
between evaluations, avoiding unnecessary object creation and
improving performance.

* Source/WebCore/css/query/MediaQueryEvaluator.cpp:
(WebCore::MQ::MediaQueryEvaluator::evaluate const):
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::invalidateMediaQueryDefaultStyle):
(WebCore::Document::mediaQueryDefaultStyle const):
(WebCore::Document::setCompatibilityMode):
* Source/WebCore/dom/Document.h:
* Source/WebCore/dom/Page.cpp:
(WebCore::Page::setNeedsRecalcStyleInAllFrames):

darinadler commented on Source/WebCore/dom/Document.h:
```diff
@@ -2375,7 +2378,9 @@ class Document
     RefPtr<MediaQueryMatcher> m_mediaQueryMatcher;
-    
+
+    mutable std::unique_ptr<RenderStyle> m_cachedMediaQueryDefaultStyle;
```
Should we store this state in the document's media query matcher instead? If we did that, we could avoid moving this code out into Document, which makes the huge class even bigger.

mstyura replied:
I've tried going down that path (though I haven't fully finished it). It seems like it would require changes in more places. The state can be moved into MediaQueryMatcher, but due to the ownership relationships between MediaQueryMatcher, Document, and MediaQueryEvaluator, changes to Document.h/cpp are unavoidable — unless an even deeper refactoring is done.

If you believe this is the more appropriate way to implement the change I want to land, let me know and I'll complete it and submit it as an alternative PR.

darinadler replied:
I do think it can be refactored to put less of this into `Document`. I understand that there will still be some code in `Document`. Long term, it’s important that we don’t just make these primary classes like `Document` grow.

I suspect an even better fix will be a refactor that eliminates the use of `RenderStyle` to pass font sizes in, which should be even more effective than reusing a `RenderStyle`.

I understand if you want to land it exactly like this, and it’s OK with me, but there’s an opportunity to do even better.

anttijk commented on Source/WebCore/css/query/MediaQueryEvaluator.cpp:
```diff
@@ -81,16 +79,7 @@ bool MediaQueryEvaluator::evaluate(const MediaQuery& query) const
-        }
-
-        FeatureEvaluationContext context { *document, { *m_rootElementStyle, &defaultStyle, nullptr, document->renderView() }, nullptr };
+        FeatureEvaluationContext context { *document, { *m_rootElementStyle, &document->mediaQueryDefaultStyle(), nullptr, document->renderView() }, nullptr };
```
@darinadler pointed out that no tests fail if you simply pass a nullptr instead of `&defaultStyle` here (`CSSToLengthConversionData::m_rootStyle`). I think we could either do that, or even better pass `m_rootElementStyle` as the second parameter as well. The `m_rootElementStyle` is the actual root element style that we want to resolve root relative units (like `rem`) against. 

(nullptr works because the unit resolution falls back to the main style if the root style field is null)

emilio replied:
@anttijk `m_rootElementStyle` doesn't seem like what you want, that causes cycles, right? As in, if a media query evaluated based on `m_rootElementStyle` but then contained `:root { font-size: ... }` inside it'd trivially be cyclic.

anttijk replied:
@Emilio I don't think so? During style resolution `m_rootElementStyle` is not actually the root element style but a special style build for the root element by matching the UA sheet only. Thus author style can't cause cycles.

darinadler replied:
Not to constantly beat the “tests” drum but:
1. If we are worried about cycles with `:root { font-size ... }` the best way to prove they don’t exist is to make sure we have a test.
2. If we think we need to pass `m_rootElementStyle` instead of `nullptr` to make `rem` work properly in media queries the best way to demonstrate that is also probably a test.

emilio replied:
@anttijk Alright, that seems fine then, I was not aware of that, was just commenting based on the:

> The m_rootElementStyle is the actual root element style :)

in your comment.

anttijk replied:
@emilio I suppose that wasn't exactly accurate!

anttijk replied:
> Not to constantly beat the “tests” drum but:
> 
> 1. If we are worried about cycles with `:root { font-size ... }` the best way to prove they don’t exist is to make sure we have a test.
> 2. If we think we need to pass `m_rootElementStyle` instead of `nullptr` to make `rem` work properly in media queries the best way to demonstrate that is also probably a test.

I think this is at least somewhat covered by the existing tests like LayoutTests/imported/w3c/web-platform-tests/css/mediaqueries/relative-units-001.html which verifies that `rem` unit in a media query is not affected by the author style.
